com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java

```diff
@@ -130,6 +146,55 @@ public void initialize(IDebugAdapterContext context) {
         }
     }
 
+    private void registerBreakpointHandler(IDebugAdapterContext context) {
+        IDebugSession debugSession = context.getDebugSession();
+        if (debugSession != null) {
+            debugSession.getEventHub().events().subscribe(debugEvent -> {
+                if (!(debugEvent.event instanceof BreakpointEvent)) {
+                    return;
+                }
+                Event event = debugEvent.event;
+                if (debugEvent.eventSet.size() > 1 && debugEvent.eventSet.stream().anyMatch(t -> t instanceof StepEvent)) {
+                    // The StepEvent and BreakpointEvent are grouped in the same event set only if they occurs at the same location and in the same thread.
+                    // In order to avoid two duplicated StoppedEvents, the debugger will skip the BreakpointEvent.
+                } else {
+                    ThreadReference bpThread = ((BreakpointEvent) event).thread();
+                    IEvaluationProvider engine = context.getProvider(IEvaluationProvider.class);
+                    if (engine.isInEvaluation(bpThread)) {
+                        return;
```

Comment:
If early return, what is the expected value of debugEvent.shouldResume?

Reply:
shouldResume is default true, that is expected.